proposals/NNNN-DXIL-Scalarization.md

```diff
@@ -0,0 +1,239 @@
+ The goal of this proposal is the present a solution for LLVM's pass manager.
+
+## Motivation
+Without Scalarizing the data structures and call instructions we can't generate valid DXIL.
+
+ ## Background
+The rewrite functions specified in the appendix cover the behavior we need to 
+support for scalarization. The specific instructions DXC operates on are
+`CallInst`, `LoadInst`, `StoreInst`, `GetElementPtrInst`, `AddrSpaceCastInst`, `BitCastInst`, 
+and `MemIntrinsic`. DXC  works on two sets of data globals\allocas and 
+function arguments via the `RewriteForScalarRepl`.
+
+There are a few ways we could tackle this work that gets us towards behavioral parity.
+A good example is to look at:
+- [SPIRVEmitIntrinsics.cpp](https://github.com/llvm/llvm-project/blob/main/llvm/lib/Target/SPIRV/SPIRVEmitIntrinsics.cpp)
+- [SROA.cpp](https://github.com/llvm/llvm-project/blob/main/llvm/lib/Transforms/Scalar/SROA.cpp)
+- [Scalarizer.cpp](https://github.com/llvm/llvm-project/blob/main/llvm/lib/Transforms/Scalar/Scalarizer.cpp)
+
+Given how the existing passes are written we should get `GetElementPtrInst`, `AddrSpaceCastInst`, `BitCastInst`, 
+and `MemIntrinsic` support mostly for free by tapping into functions like 
+`visitGetElementPtrInst`, `visitBitCastInst`,  `visitMemSetInst`, & `visitMemTransferInst`.
+
+There likely shouldn't be any need for `AddrSpaceCastInst`. 
+
+We also will get `LoadInst` and `StoreInst` for non DXIL resources  via  `visitLoadInst` and `visitStoreInst`.
+
+Finally the `CallInst` cases is really just to cover HLSL intrinsics. This can be handled via a simple pass
+that knows all the intrinsics and loops through the vector type to emit the correct number of calls per element.
+
+## Priority of work
+- A good starting place of behaviors to try to match is `RewriteCall` 
+specifically `RewriteWithFlattenedHLIntrinsicCall`As the other pieces are not needed 
+for compute shaders. 
+- Next would be `RewriteForLoad` and `RewriteForStore` as those 
+transformations are need to properly handle resources and breakdown to 
+three simple cases.
+
+The behaviors of `RewriteForGEP`, `RewriteForAddrSpaceCast`, `RewriteForConstExpr`,
+and `RewriteMemIntrin`. Should be things we can get  via existing LLVM passes.
+
+## Areas explicitly ignored
+SROA_Parameter pass has some  matrix loads and stores that this proposal won't go into.
+It also looks to do some legalization via `LegalizeDxilInputOutputs` that won't be covered.
+
+## Tickets
+1. Create a scalarization pass that just handles the flattenArgument case to start.      
+    - To simplify things "workists" will just be HLSL intrinsics.
+    - We iterate over all intrinsics and iterate over all value uses
+    - For now the only value we will check for are `CallInst`
+    - The pass will replace intrinsics used by DXIL.td that are a multi-element 
+    call with multiple scalarized calls
+2. Expand the scalar expansion pass to handle `loads` by checking for `LoadInst`
```

Comment:
How does this align with the resources plan (@bogner)? Are we going to have custom load instructions early in the compiler or late?